fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/query/FDBInQueryTest.java

```diff
@@ -2553,4 +2558,123 @@ void testInQueryWithNonSargableReplan(final boolean shouldReplan) throws Excepti
+                .setFilter(inFilter)
+                .setSort(field("num_value_3_indexed"))
+                .build();
+
+        RecordQueryPlan plan = planQuery(query);
+
+        // Make sure we do not plan the IN as a union and that we don't fall back to a primary scan but use an
+        // index.
+        if (useCascadesPlanner) {
+            assertMatchesExactly(plan,
+                    predicatesFilterPlan(
+                            indexPlan()
+                                    .where(indexName("MySimpleRecord$num_value_3_indexed"))
+                                    .and(scanComparisons(unbounded())))
+                            .where(predicates(only(valuePredicate(anyValue(), new Comparisons.ListComparison(Comparisons.Type.IN, List.of(1, 3, 5, 7, 9)))))));
+            assertEquals(-441605751, plan.planHash(PlanHashable.CURRENT_LEGACY));
+            assertEquals(1552275015, plan.planHash(PlanHashable.CURRENT_FOR_CONTINUATION));
+        } else {
+            assertMatchesExactly(plan,
+                    filterPlan(
+                            indexPlan()
+                                    .where(indexName("MySimpleRecord$num_value_3_indexed"))
+                                    .and(scanComparisons(unbounded())))
+                            .where(queryComponents(only(equalsObject(inFilter)))));
+            assertEquals(1975903190, plan.planHash(PlanHashable.CURRENT_LEGACY));
+            assertEquals(1012163023, plan.planHash(PlanHashable.CURRENT_FOR_CONTINUATION));
+        }
+    }
+
+    @DualPlannerTest
+    @ParameterizedTest(name = "testInUnionWithSomeSargable[replans={0}]")
+    @ValueSource(ints = {-2, -1, 0, 1})
+    void testInUnionWithSomeSargable(int replans) throws Exception {
+        final Index numValue2Then3Index = new Index("numValue2NumValue3", concatenateFields("num_value_2", "num_value_3_indexed"));
+        final RecordMetaDataHook hook = metaDataBuilder -> metaDataBuilder.addIndex("MySimpleRecord", numValue2Then3Index);
+        complexQuerySetup(hook);
+
+        var plannerConfigBuilder = planner.getConfiguration().asBuilder()
+                .setAttemptFailedInJoinAsUnionMaxSize(100)
+                .setOmitPrimaryKeyInOrderingKeyForInUnion(true);
+        if (replans != -1) {
+            plannerConfigBuilder.setMaxNumReplansForInUnion(replans);
+        }
+        planner.setConfiguration(plannerConfigBuilder.build());
+        final var inFilter1 = Query.field("num_value_2").in("nv2_list");
+        final var inFilter2 = Query.field("str_value_indexed").in("str_list");
+        RecordQuery query = RecordQuery.newBuilder()
+                .setRecordType("MySimpleRecord")
+                .setFilter(Query.and(inFilter1, inFilter2))
+                .setSort(field("num_value_3_indexed"))
+                .build();
+
+        RecordQueryPlan plan = planQuery(query);
+
+        if (useCascadesPlanner) {
+            // This is a suboptimal plan!
```

Comment:
Note that the Cascades planner doesn't quite get this right. (There are a few other plans in this fixture that have similar issues.) I left them as is for now, but we could tackle it. I think the issue is that the cost model looks at the number of non-sargable in comparisons when evaluating plans, but it can be tricked if the offending plan is nested.

Reply:
That is unfortunately correct. Can you open an issue please?

Reply:
Filed #2950